src/main/java/net/neoforged/neoforge/common/util/NeoForgeExtraCodecs.java

```diff
@@ -112,11 +117,28 @@ public static <A> Codec<List<A>> listWithoutEmpty(Codec<List<Optional<A>>> codec
      * the first codec and then the second codec for decoding, <b>but only the first for encoding</b>.
      * <p>
      * Unlike vanilla, this alternative codec also tries to encode with the second codec if the first encode fails.
+     * <p>
+     * This codec will also try to keep the resulting codec to a MapCodecCodec if both inputs are MapCodecCodecs.
      */
     public static <T> Codec<T> withAlternative(final Codec<T> codec, final Codec<T> alternative) {
+        if (codec instanceof MapCodec.MapCodecCodec<T> mCodec && alternative instanceof MapCodec.MapCodecCodec<T> mAlternative) {
+            //If both codecs are MapCodecCodecs then return it as a MapCodecCodec to allow for cleaner support in dispatch codecs
+            return withAlternative(mCodec.codec(), mAlternative.codec()).codec();
+        }
```

Comment:
This goes against the typical design of codecs: it is the user's responsibility to ensure that `MapCodec`s get turned into `Codec`s at the last moment.

Reply:
Ah yeah I guess I can see how `ExtraCodecs#validate` is a special case for why it checks explicitly for if it is a `MapCodecCodec`. As they don't want to add in extra overhead/logic when it comes to converting certain bits to being validated and accidentally causing them to no longer be `MapCodecCodecs`